### PR TITLE
Add react native docs update script

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This helps us maintain the React Native (API) Docs.

# How

It uses the submodule, an Expo SDK version (`--sdk`) and commit range (`--from` and optionally `--to`) to check for added, removed or updated files.
- All added files are prefixed with `ADDED_*.md` and ignored from the repo 
- All removed files are prefixed with `REMOVED_*.md` and ignored from the repo
- All updated files receive a `.diff` to show the developer what has changed

With this, we can manually apply changes and/or modify the relevant files from our docs. It also can diverge from the original React Native repository without causing "too much" trouble.

# Test Plan

Here is a test command:

```
$ et update-react-native-docs --sdk unversioned --from 9806dddd7b13df2f9d7b12bfd8255338640caf88 
```
> This will update the **unversioned** folder with changes between `9806ddd...master`

### CODE REVIEW

**Review by tsapeta — COMMENTED** [review]
I haven't reviewed the code yet, but there is `--update-react-native-docs` flag in `et generate-sdk-docs` command so we might want to remove this flag from there if this new command is better. Could you also update https://github.com/expo/expo/blob/master/guides/releasing/Release%20Workflow.md#63-generate-and-deploy-new-docs accordingly?

**Comment by byCedric** [comment]
Yes definitely, I first wanted to do a code review with Brent. After green light, I'll update the rest. But, to shortly reiterate on some differences with `et generate-sdk-docs`:

- This script "adds" and "removes" files based on the differences with upstream, using the relevant documentation areas.
- This script can update existing docs, it will show the dev a `.diff` for each changed file.
- This script doesn't use the "old" `import-react-native-docs`. Instead, it relies on some more manual steps in order to make the React Native API Docs better in the long run.

File: tools/expotools/src/commands/UpdateReactNativeDocs.ts [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+  }
+
+  await updateDocsAsync(options);
+
+  const summary = getDocsSummary(
+    await getLocalFilesAsync(options),
+    await getUpstreamFilesAsync(options),
+  );
+
+  console.log('');
```
Comment (tsapeta) [inline tools/expotools/src/commands/UpdateReactNativeDocs.ts:40]: `console.log()` without params works too 😉

File: tools/expotools/src/commands/UpdateReactNativeDocs.ts [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  sdk: string;
+  from: string;
+  to: string;
+}
+
+interface DocsSummary {
+  removed: string[];
+  added: string[];
+  changed: string[];
+}
```
Comment (tsapeta) [inline tools/expotools/src/commands/UpdateReactNativeDocs.ts:242]: We usually put types/interfaces on top of the file. Also, they should probably be changed to types — as they are not intended to be implemented by classes 🙂

File: tools/expotools/src/commands/UpdateReactNativeDocs.ts [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+const PREFIX_ADDED = 'ADDED_';
+const PREFIX_REMOVED = 'REMOVED_';
+
+const DOCS_IGNORED = [
+  'appregistry',
+  'components-and-apis',
+  'drawerlayoutandroid',
+  'linking',
+  'settings',
+  'settings',
```
Comment (tsapeta) [inline tools/expotools/src/commands/UpdateReactNativeDocs.ts:22]:
```suggestion
  'settings',
```

File: tools/expotools/src/commands/UpdateReactNativeDocs.ts [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+  }
+
+  for (const entry of summary.removed) {
+    if (entry.startsWith(PREFIX_REMOVED)) {
+      continue;
+    }
+
+    await fs.move(
+      path.join(DOCS_DIR, 'pages', 'versions', options.sdk, 'react-native', `${entry}.md`),
+      path.join(DOCS_DIR, 'pages', 'versions', options.sdk, 'react-native', `${PREFIX_REMOVED}${entry}.md`),
```
Comment (tsapeta) [inline tools/expotools/src/commands/UpdateReactNativeDocs.ts:190]: I think it's worth to join this directory once and then join with just filenames — easier to read, easier to maintain

**Review by tsapeta — COMMENTED** [review]
Looks good so far 👍 

I'm wondering whether we could parallelize some of those `for` loops with `await` inside, using `Promise.all` 🤔 Those operations seem to be quick so that's fine, but in general I recommend using it for longer tasks.

Also, I've recently added `Logger` class to expotools which uses some chalks depending on the log level, you may find it useful but I'm ok with the current solution 😉

**Comment by byCedric** [comment]
Thanks for the review, I'll iterate on that asap 😄

**Comment by byCedric** [comment]
Done, all feedback is in there I think 😄

File: tools/expotools/src/commands/UpdateReactNativeDocs.ts [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+  logger.log();
+
+  return useDirtyGit;
+}
+
+async function updateDocsAsync(options: Options) {
+  logger.info(`📚 Updating ${chalk.cyan('react-native-website')} submodule...`);
+
+  await spawnAsync('git', ['checkout', 'master'], { cwd: RN_REPO_DIR });
```
Comment (tsapeta) [inline tools/expotools/src/commands/UpdateReactNativeDocs.ts:132]: Sorry that I haven't noticed this before – there is also a tool in `Git.ts` that you can use here 😅 
Just import `GitDirectory` from there and create an instance like this `new GitDirectory(RN_REPO_DIR);`.
Probably some commands needs to be added there (`checkout` for example) but it's worth to migrate it to this tool 😉

Comment (byCedric) [reply tools/expotools/src/commands/UpdateReactNativeDocs.ts:132]: Sure, sounds good! I'll add that, thanks for the tip.

Comment (byCedric) [reply tools/expotools/src/commands/UpdateReactNativeDocs.ts]: Done! Works pretty good 😄

Comment (tsapeta) [reply tools/expotools/src/commands/UpdateReactNativeDocs.ts]: Yeah, sure! Sounds good, it's been an oversight to require it 😅

**Review by tsapeta — APPROVED** [review]

File: tools/expotools/src/commands/UpdateReactNativeDocs.ts [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+
+const RN_REPO_DIR = path.join(DOCS_DIR, 'react-native-website');
+const RN_WEBSITE_DIR = path.join(RN_REPO_DIR, 'website');
+const RN_DOCS_DIR = path.join(RN_REPO_DIR, 'docs');
+
+const PREFIX_ADDED = 'ADDED_';
+const PREFIX_REMOVED = 'REMOVED_';
+const SUFFIX_CHANGED = '.diff';
+
+const DOCS_IGNORED = [
```
Comment (brentvatne) [inline tools/expotools/src/commands/UpdateReactNativeDocs.ts:34]: can we generate this from looking at our docs directory? so if a file existed in both `from` and `to` but does not exist in the docs directory then we consider it ignored.

don't have a strong opinion about it, but the benefit would be that we wouldn't have to manually maintain a list

Comment (byCedric) [reply tools/expotools/src/commands/UpdateReactNativeDocs.ts:34]: Yeah, that might be possible. I have to check the exact format patches, but basically:

> If diff adds file
>   - new file added (not ignored)
> else
>   - file is ignored

Comment (brentvatne) [reply tools/expotools/src/commands/UpdateReactNativeDocs.ts:34]: i think it's not a blocker for this pr but something to follow up with on afterwards maybe

**Review by brentvatne — APPROVED** [review]
thanks for the review @tsapeta!

File: tools/expotools/src/commands/UpdateReactNativeDocs.ts (hunk 6 above)
Comment (byCedric) [reply tools/expotools/src/commands/UpdateReactNativeDocs.ts:34]: Ok, let's merge this one for now to at least unblock the RN versions update. I'll add more PRs fine-tuning the behavior with the tips from this PR! The 3 PRs in addition I'll be working on:
- [ ] Tweak the Expo Tools Github wrapper to allow empty options
- [ ] Auto-detect ignored files from updated version range
- [ ] Optionally, tweak the github range input with cursors

**Pull request merged**